File: other-platforms/flutter/metakeep_flutter_sdk/android/src/main/kotlin/xyz/metakeep/sdk/metakeep_flutter_sdk/MetakeepFlutterSdkPlugin.kt
package xyz.metakeep.sdk.metakeep_flutter_sdk

import android.app.Activity
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import org.json.JSONArray
import org.json.JSONObject
import xyz.metakeep.sdk.*

/** MetaKeepFlutterSdkPlugin */
class MetaKeepFlutterSdkPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    // / The MethodChannel that will the communication between Flutter and native Android
    // /
    // / This local reference serves to register the plugin with the Flutter Engine and unregister it
    // / when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "metakeep_flutter_sdk")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        currentActivity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        TODO("Not yet implemented")
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        TODO("Not yet implemented")
    }

    override fun onDetachedFromActivity() {
        TODO("Not yet implemented")
    }

    override fun onMethodCall(
        call: MethodCall,
        result: Result,
    ) {
        when (call.method) {
            INITIALIZE_METHOD -> initialize(call, result)
            SET_USER_METHOD -> setUser(call, result)
            SIGN_MESSAGE_METHOD -> signMessage(call, result)
            SIGN_TRANSACTION_METHOD -> signTransaction(call, result)
            SIGN_TYPED_DATA_METHOD -> signTypedData(call, result)
            GET_CONSENT_METHOD -> getConsent(call, result)
            GET_WALLET_METHOD -> getWallet(call, result)
            else -> result.notImplemented()
        }
    }

    private fun initialize(
        call: MethodCall,
        result: Result,
    ) {
        call.argument<String>(APP_ID_FIELD)?.let { appId ->
            sdk = MetaKeep(appId, AppContext(currentActivity!!))
            result.success(null)
        } ?: run {
            rejectWithErrorStatus(
                INVALID_ARGUMENTS_ERROR_STATUS,
                result,
            )
        }
    }

    private fun setUser(
        call: MethodCall,
        result: Result,
    ) {
        // Check user field and then email field in the user object
        val user = call.argument<Map<String, String>>(USER_FIELD)

        if (user == null) {
            rejectWithErrorStatus(
                INVALID_USER_ERROR_STATUS,
                result,
            )
            return
        }

        val email = user[EMAIL_FIELD]

        if (email == null) {
            rejectWithErrorStatus(
                INVALID_ARGUMENTS_ERROR_STATUS,
                result,
            )
            return
        }

        sdk.user = User(email)
    }

    private fun signMessage(
        call: MethodCall,
        result: Result,
    ) {
        val message = call.argument<String>(MESSAGE_FIELD)
        val reason = call.argument<String>(REASON_FIELD)

        if (message == null || reason == null) {
            rejectWithErrorStatus(
                INVALID_ARGUMENTS_ERROR_STATUS,
                result,
            )
            return
        }

        sdk.signMessage(
            message,
            reason,
            getCallback(result),
        )
    }

    private fun signTransaction(
        call: MethodCall,
        result: Result,
    ) {
        val transaction =
            call.argument<
                HashMap<String, Any>,
                >(TRANSACTION_FIELD)
        val reason = call.argument<String>(REASON_FIELD)

        if (transaction == null || reason == null) {
            rejectWithErrorStatus(
                INVALID_ARGUMENTS_ERROR_STATUS,
                result,
            )
            return
        }

        sdk.signTransaction(
            JsonRequest(
                JSONObject(transaction as Map<*, *>?).toString(),
            ),
            reason,
            getCallback(result),
        )
    }

    private fun signTypedData(
        call: MethodCall,
        result: Result,
    ) {
        val typedData =
            call.argument<
                HashMap<String, Any>,
                >(TYPED_DATA_FIELD)
        val reason = call.argument<String>(REASON_FIELD)

        if (typedData == null || reason == null) {
            rejectWithErrorStatus(
                INVALID_ARGUMENTS_ERROR_STATUS,
                result,
            )
            return
        }

        sdk.signTypedData(
            JsonRequest(JSONObject(typedData as Map<*, *>?).toString()),
            reason,
            getCallback(result),
        )
    }

    private fun getConsent(
        call: MethodCall,
        result: Result,
    ) {
        val consentToken = call.argument<String>(CONSENT_TOKEN_FIELD)

        if (consentToken == null) {
            rejectWithErrorStatus(
                INVALID_ARGUMENTS_ERROR_STATUS,
                result,
            )
            return
        }

        sdk.getConsent(
            consentToken,
            getCallback(result),
        )
    }

    private fun getWallet(
        call: MethodCall,
        result: Result,
    ) {
        sdk.getWallet(
            getCallback(result),
        )
    }

    private fun getCallback(result: Result): Callback {
        return Callback(
            onSuccess = { response: JsonResponse ->
                result.success(
                    // Parse JSON data as dynamic map
                    JSONObject(response.data.toString()).toMap(),
                )
            },
            // Pack the actual JSON error response into the error details.
            // This can be extracted in the Flutter side.
            onFailure = { error: JsonResponse ->
                result.error(
                    OPERATION_FAILED_ERROR_STATUS,
                    OPERATION_FAILED_ERROR_STRING,
                    JSONObject(error.data.toString()).toMap(),
                )
            },
        )
    }

    private fun rejectWithErrorStatus(
        errorStatus: String,
        result: Result,
    ) {
        result.error(
            OPERATION_FAILED_ERROR_STATUS,
            OPERATION_FAILED_ERROR_STRING,
            mapOf(STATUS_FIELD to errorStatus),
        )
    }

    private fun JSONObject.toMap(): Map<String, *> =
        keys().asSequence().associateWith { it ->
            when (val value = this[it]) {
                is JSONArray ->
                    {
                        val map = (0 until value.length()).associate { Pair(it.toString(), value[it]) }
                        JSONObject(map).toMap().values.toList()
                    }
                is JSONObject -> value.toMap()
                JSONObject.NULL -> null
                else -> value
            }
        }

    // Holds the MetaKeep SDK instance
    private lateinit var sdk: MetaKeep

    // Holds the current activity
    private lateinit var currentActivity: Activity

    companion object {
        const val NAME = "MetaKeepReactNativeSDK"

        // Methods
        const val INITIALIZE_METHOD = "initialize"
        const val SET_USER_METHOD = "setUser"
        const val SIGN_MESSAGE_METHOD = "signMessage"
        const val SIGN_TRANSACTION_METHOD = "signTransaction"
        const val SIGN_TYPED_DATA_METHOD = "signTypedData"
        const val GET_CONSENT_METHOD = "getConsent"
        const val GET_WALLET_METHOD = "getWallet"

        // Error status
        const val INVALID_ARGUMENTS_ERROR_STATUS = "INVALID_ARGUMENTS"
        const val INVALID_USER_ERROR_STATUS = "INVALID_USER"
        const val OPERATION_FAILED_ERROR_STATUS = "OPERATION_FAILED"

        // Error strings
        const val OPERATION_FAILED_ERROR_STRING = "MetaKeep SDK operation failed"

        // Constant strings
        const val APP_ID_FIELD = "appId"
        const val USER_FIELD = "user"
        const val EMAIL_FIELD = "email"
        const val MESSAGE_FIELD = "message"
        const val TRANSACTION_FIELD = "transaction"
        const val TYPED_DATA_FIELD = "typedData"
        const val REASON_FIELD = "reason"
        const val STATUS_FIELD = "status"
        const val CONSENT_TOKEN_FIELD = "consentToken"
    }
}

File: other-platforms/flutter/metakeep_flutter_sdk/ios/Classes/MetaKeepFlutterSdkPlugin.swift
import Flutter
import MetaKeep
import UIKit

public class MetaKeepFlutterSdkPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "metakeep_flutter_sdk", binaryMessenger: registrar.messenger())
    let instance = MetaKeepFlutterSdkPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case INITIALIZE_METHOD:
      return initialize(call, result: result)
    case SET_USER_METHOD:
      return setUser(call, result: result)
    case SIGN_MESSAGE_METHOD:
      return signMessage(call, result: result)
    case SIGN_TRANSACTION_METHOD:
      return signTransaction(call, result: result)
    case SIGN_TYPED_DATA_METHOD:
      return signTypedData(call, result: result)
    case GET_CONSENT_METHOD:
      return getConsent(call, result: result)
    case GET_WALLET_METHOD:
      return getWallet(call, result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  func initialize(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
      let appId = args[APP_ID_FIELD] as? String
    else {
      return rejectWithErrorStatus(errorStatus: INVALID_ARGUMENTS_ERROR_STATUS, result: result)
    }
    sdk = MetaKeep(appId: appId, appContext: AppContext())
    result(nil)
  }

  func setUser(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    // Only email is supported for now
    guard let args = call.arguments as? [String: Any],
      let user = args[USER_FIELD] as? [String: Any],
      let email = user[EMAIL_FIELD] as? String
    else {
      return rejectWithErrorStatus(errorStatus: INVALID_USER_ERROR_STATUS, result: result)
    }

    sdk!.user =
      User(
        email: email
      )
  }

  func signMessage(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
      let message = args[MESSAGE_FIELD] as? String,
      let reason = args[REASON_FIELD] as? String
    else {
      return rejectWithErrorStatus(errorStatus: INVALID_ARGUMENTS_ERROR_STATUS, result: result)
    }

    sdk!.signMessage(
      message: message,
      reason: reason,
      callback: getCallback(result))
  }

  func signTransaction(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
      let transaction = args[TRANSACTION_FIELD] as? NSDictionary,
      let reason = args[REASON_FIELD] as? String
    else {
      return rejectWithErrorStatus(errorStatus: INVALID_ARGUMENTS_ERROR_STATUS, result: result)
    }

    do {
      sdk!.signTransaction(
        transaction: try NSDictionaryToJsonRequest(dictionary: transaction as NSDictionary),
        reason: reason,
        callback: getCallback(result))
    } catch {
      rejectWithErrorStatus(errorStatus: INVALID_TRANSACTION_ERROR_STATUS, result: result)
    }
  }

  func signTypedData(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
      let typedData = args[TYPED_DATA_FIELD] as? NSDictionary,
      let reason = args[REASON_FIELD] as? String
    else {
      return rejectWithErrorStatus(errorStatus: INVALID_ARGUMENTS_ERROR_STATUS, result: result)
    }

    do {
      sdk!.signTypedData(
        typedData: try NSDictionaryToJsonRequest(dictionary: typedData),
        reason: reason,
        callback: getCallback(result))
    } catch {
      rejectWithErrorStatus(errorStatus: INVALID_TYPED_DATA_ERROR_STATUS, result: result)
    }
  }

  func getConsent(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
      let consentToken = args[CONSENT_TOKEN_FIELD] as? String
    else {
      return rejectWithErrorStatus(errorStatus: INVALID_ARGUMENTS_ERROR_STATUS, result: result)
    }

    sdk!.getConsent(
      consentToken: consentToken,
      callback: getCallback(result))
  }

  func getWallet(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    sdk!.getWallet(
      callback: getCallback(result))
  }

  private func getCallback(_ result: @escaping FlutterResult) -> Callback {
    return Callback(
      onSuccess: { (response: JsonResponse) in
        result(response.data)
      },
      onFailure: { (error: JsonResponse) in
        result(
          FlutterError(
            // Pack the actual JSON error response into the error details
            // so that the caller can access it.
            code: self.OPERATION_FAILED_ERROR_STATUS, message: self.OPERATION_FAILED_ERROR_STRING,
            details: error.data
          ))
      }
    )
  }

  private func NSDictionaryToJsonRequest(dictionary: NSDictionary) throws -> JsonRequest {
    let jsonString = String(
      bytes: try JSONSerialization.data(withJSONObject: dictionary), encoding: .utf8)!
    return try JsonRequest(jsonString: jsonString)
  }

  private func rejectWithErrorStatus(errorStatus: String, result: @escaping FlutterResult) {
    result(
      FlutterError(
        code: self.OPERATION_FAILED_ERROR_STATUS, message: OPERATION_FAILED_ERROR_STRING,
        details: [STATUS_FIELD: errorStatus]
      ))
  }

  // Holds the MetaKeep SDK instance
  private var sdk: MetaKeep?

  // Methods
  private let INITIALIZE_METHOD = "initialize"
  private let SET_USER_METHOD = "setUser"
  private let SIGN_MESSAGE_METHOD = "signMessage"
  private let SIGN_TRANSACTION_METHOD = "signTransaction"
  private let SIGN_TYPED_DATA_METHOD = "signTypedData"
  private let GET_CONSENT_METHOD = "getConsent"
  private let GET_WALLET_METHOD = "getWallet"

  // Error status
  private let INVALID_ARGUMENTS_ERROR_STATUS = "INVALID_ARGUMENTS"
  private let INVALID_USER_ERROR_STATUS = "INVALID_USER"
  private let INVALID_TRANSACTION_ERROR_STATUS = "INVALID_TRANSACTION"
  private let INVALID_TYPED_DATA_ERROR_STATUS = "INVALID_TYPED_DATA"
  private let OPERATION_FAILED_ERROR_STATUS = "OPERATION_FAILED"

  // Error strings
  private let OPERATION_FAILED_ERROR_STRING = "MetaKeep SDK operation failed"

  // Constant strings
  private let METAKEEP_DOMAIN = "MetaKeep"
  private let APP_ID_FIELD = "appId"
  private let USER_FIELD = "user"
  private let EMAIL_FIELD = "email"
  private let MESSAGE_FIELD = "message"
  private let TRANSACTION_FIELD = "transaction"
  private let TYPED_DATA_FIELD = "typedData"
  private let REASON_FIELD = "reason"
  private let STATUS_FIELD = "status"
  private let CONSENT_TOKEN_FIELD = "consentToken"
}
